lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,517 @@
+    break;
+  case BTOR2_TAG_not:
+    res = builder.create<btor::NotOp>(unknownLoc, cache.at(kids[0]));
+    break;
+  case BTOR2_TAG_redand:
+    res = builder.create<btor::RedAndOp>(unknownLoc, builder.getIntegerType(1),
+                                         cache.at(kids[0]));
+    break;
+  case BTOR2_TAG_redor:
+    res = builder.create<btor::RedOrOp>(unknownLoc, builder.getIntegerType(1),
+                                        cache.at(kids[0]));
+    break;
+  case BTOR2_TAG_redxor:
+    res = builder.create<btor::RedXorOp>(unknownLoc, builder.getIntegerType(1),
+                                         cache.at(kids[0]));
+    break;
+  case BTOR2_TAG_const:
+    res = buildConstantOp(line->sort.bitvec.width,
+                        std::string(line->constant), 2);
+    break;
+  case BTOR2_TAG_constd:
+    res = buildConstantOp(line->sort.bitvec.width, 
+                        std::string(line->constant), 10);
+    break;
+  case BTOR2_TAG_consth:
+    res = buildConstantOp(line->sort.bitvec.width, 
+                        std::string(line->constant), 16);
+    break;
+  case BTOR2_TAG_one:
+    res = buildConstantOp(line->sort.bitvec.width, 
+                        std::string("one"), 10);
+    break;
+  case BTOR2_TAG_ones:
+    res = buildConstantOp(line->sort.bitvec.width,
+                        std::string("ones"), 10);
+    break;
+  case BTOR2_TAG_zero:
+    res = buildConstantOp(line->sort.bitvec.width, 
+                        std::string("zero"), 10);
+    break;
+  case BTOR2_TAG_bad:
+    res = builder.create<btor::AssertNotOp>(unknownLoc, cache.at(kids[0]));
+    break;
+
+  // indexed ops
+  case BTOR2_TAG_slice:
+    res = buildSliceOp(cache.at(kids[0]), kids[1], kids[2]);
+    break;
+  case BTOR2_TAG_sext:
+    res = builder.create<btor::SExtOp>(
+        unknownLoc, cache.at(kids[0]),
+        builder.getIntegerType(line->sort.bitvec.width));
+    break;
+  case BTOR2_TAG_uext:
+    res = builder.create<btor::UExtOp>(
+        unknownLoc, cache.at(kids[0]),
+        builder.getIntegerType(line->sort.bitvec.width));
+    break;
+
+  // ternary ops
+  case BTOR2_TAG_ite:
+    res = builder.create<btor::IteOp>(unknownLoc, cache.at(kids[0]),
+                                      cache.at(kids[1]), cache.at(kids[2]));
+    break;
+
+  // unmapped ops
+  case BTOR2_TAG_read:
+  case BTOR2_TAG_constraint:
+  case BTOR2_TAG_init:
+  case BTOR2_TAG_input:
+  case BTOR2_TAG_next:
+  case BTOR2_TAG_sort:
+  case BTOR2_TAG_state:
+  case BTOR2_TAG_fair:
+  case BTOR2_TAG_justice:
+  case BTOR2_TAG_output:
+  case BTOR2_TAG_write:
+  default:
+    break;
+  }
+  return res;
+}
+
+void Deserialize::createNegateLine(int64_t curAt, Value &child) {
+  auto res = builder.create<btor::NotOp>(unknownLoc, cache.at(curAt * -1));
+  assert(res && res->getNumResults() == 1);
+  cache[curAt] = res->getResult(0);
+}
+
+bool Deserialize::isValidChild(Btor2Line * line) {
+  auto tag = reachedLines.at(line->id)->tag;
```

Comment:
1. In other code, we usually use a convention to name class fields with prefix `m_`.  This makes it easier to read code since it is clear what is a field and what is local.
2. I would have called `reachedLines` simply `m_lines`. Less typing, no multiple meanings for **reachable**
3. I would wrap access to fields with functions. Makes code easier to read and easier to debug (by adding assertions). For example, you clearly need `lineById(uint32_t id)` function.

Reply:
Got it!